### PR TITLE
add resync logic for addon

### PR DESCRIPTION
**Pull request by elgnay**
Signed-off-by: Yang Le <yangle@redhat.com>

### CODE REVIEW

File: pkg/spoke/addon/registration_controller.go [hunk 1]
```diff
@@ -74,11 +79,44 @@ func NewAddOnRegistrationController(
...
+
+	// handle resync
+	addOns, err := c.hubAddOnLister.ManagedClusterAddOns(c.clusterName).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	errs := []error{}
+	for _, addOn := range addOns {
+		_, err := c.hubAddOnClient.AddonV1alpha1().ManagedClusterAddOns(c.clusterName).Get(ctx, addOn.Name, metav1.GetOptions{})
```
Comment (qiujian16) [inline pkg/spoke/addon/registration_controller.go]: why not get from cache?

File: pkg/spoke/addon/registration_controller.go [hunk 2]
```diff
@@ -74,11 +79,44 @@ func NewAddOnRegistrationController(
...
 func (c *addOnRegistrationController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
-	addOnName := syncCtx.QueueKey()
+	queueKey := syncCtx.QueueKey()
+	if queueKey != factory.DefaultQueueKey {
+		// sync a particular addOn
+		return c.syncAddOn(ctx, syncCtx, queueKey)
+	}
+
+	// handle resync
+	addOns, err := c.hubAddOnLister.ManagedClusterAddOns(c.clusterName).List(labels.Everything())
```
Comment (qiujian16) [inline pkg/spoke/addon/registration_controller.go]: i think we get all existing addons from addOnRegistrationConfigs,  so we can compare with current lister

**Review by qiujian16 — COMMENTED** [review]
we  need some test for this

File: pkg/spoke/addon/registration_controller.go (hunk 1 above)
Comment (elgnay) [reply pkg/spoke/addon/registration_controller.go]: Code updated to sync between cached registration configs and ManagedClusterAddOns in informer cache.

File: pkg/spoke/addon/registration_controller.go (hunk 2 above)
Comment (elgnay) [reply pkg/spoke/addon/registration_controller.go]: Code updated to sync between cached registration configs and ManagedClusterAddOns in informer cache.

**Comment by elgnay** [comment]
> we need some test for this

Unit test case added.

**Pull request merged**